add how-to create a chisseled ubuntu image and use it to launch c/c++… (#36)
canonical/ubuntu-cloud-docs: oci/oci-how-to/create-chiselled-ubuntu-image.rst

Create a chiselled Ubuntu base image for C/C++, Go and Rust applications
========================================================================

> **This guide will provide step-by-step instructions on how  to create your own chiselled Ubuntu container image to run a compiled application**

Reviewer: ```suggestion
This guide will provide step-by-step instructions on how  to create your own chiselled Ubuntu container image to run a compiled application.
```

> - Chiselled Ubuntu containers combine Distroless and Ubuntu to create smaller, more secure containers.

Reviewer 2: ```suggestion
- Chiselled Ubuntu are appliance-type container images combining both the advantages of Distroless and Ubuntu to create smaller, more secure containers, without loosing the value add of a stable Linux distribution.
```

> - **Chiselled Ubuntu containers combine Distroless and Ubuntu to create smaller, more secure containers.**

Reviewer: ```suggestion
- Chiselled Ubuntu containers combine Distroless and Ubuntu to create smaller, more secure containers.
```

> - **The reduced size of the containers reduces the overall attack surface. Combined with the support and content quality from the Ubuntu distribution, chiselled Ubuntu is a significant security improvement.**

Reviewer: ```suggestion
- The reduced size of the containers reduces the overall attack surface. Combined with the support and content quality from the Ubuntu distribution, chiselled Ubuntu is a significant security improvement.
```

> - **Chisel provides a developer-friendly CLI to install slices of packages from the upstream Ubuntu distribution onto container filesystem.**

Reviewer: ```suggestion
- Chisel provides a developer-friendly CLI to install slices of packages from the upstream Ubuntu distribution onto the container filesystem.
```

Build the chiselled Ubuntu base image
-------------------------------------

This image must contain all the essential slices that are required for the execution of common compiled applications. As a base, the must-have list of slices is:

> - **base-files_base and base-files_release-info**: will give you the overall base structure of the container image's filesystem, together with the underlying Ubuntu base's release information

Reviewer: ```suggestion
- **base-files_base** and **base-files_release-info**: will give you the overall base structure of the container image's filesystem, together with the underlying Ubuntu base's release information,
```

> - **ca-certificates_data**: for cryptographic certificate verifications

Reviewer: ```suggestion
- **ca-certificates_data**: for cryptographic certificate verifications,
```

> - **libgcc-s1_libs** and **libc6_libs**: for the libgcc-s1 and libc6 libraries

Reviewer: ```suggestion
- **libgcc-s1_libs** and **libc6_libs**: for the libgcc-s1 and libc6 libraries.
```

   FROM golang:1.18 AS chisel
   ARG UBUNTU_RELEASE
   RUN git clone --depth 1 -b main https://github.com/canonical/chisel /opt/chisel
   WORKDIR /opt/chisel
   RUN go generate internal/deb/version.go \
      && go build ./cmd/chisel

   # Build the chiselled filesystem based on the desired slices.
   FROM ubuntu:$UBUNTU_RELEASE AS builder
   ARG UBUNTU_RELEASE
   RUN apt-get update \
      && DEBIAN_FRONTEND=noninteractive apt-get install -y ca-certificates
   COPY --from=chisel /opt/chisel/chisel /usr/bin/
   WORKDIR /rootfs
   RUN chisel cut --release ubuntu-$UBUNTU_RELEASE --root /rootfs \
      base-files_base \
      base-files_release-info \
      ca-certificates_data \
      libgcc-s1_libs \
      libc6_libs  

   # Make the chiselled filesystem the only thing
   # present in the final image.
   FROM scratch
   COPY --from=builder /rootfs /

Build the chiselled Ubuntu base image by running:

..  code-block:: bash
   
   docker build -t chiselled-ubuntu-base:latest .

You'll then find yourself with a new container image with approximately 5MB (or 2.5MB when compressed). 

Build the application image
---------------------------

> Now that you have the `chiselled-ubuntu-base:latest` image, you can simply add your compiled application to the image and run it from there. For the sake of simplicity, this guide will give your three very simple “Hello World” application examples for C, Go and Rust.

Reviewer: ```suggestion
Now that you have the ``chiselled-ubuntu-base:latest`` image, you can simply add your compiled application to the image and run it from there. For the sake of simplicity, this guide will give your three very simple “Hello World” application examples for C, Go and Rust.
```

.. tabs::

	.. tab:: C

>          Assume the following `app.c` source code:

Reviewer: ```suggestion
         Assume the following ``app.c`` source code:
```

         .. code-block:: c

            #include <stdio.h>

            int main() {
               printf("Hello World!");
               return 0;
            }
            

>          You'd typically compile this via `gcc app.c -o app`.

Reviewer: ```suggestion
         You'd typically compile this via ``gcc app.c -o app``.
```

   
	.. tab:: Go

>          Assume the following `app.go` source code:

Reviewer: ```suggestion
         Assume the following ``app.go`` source code:
```

         .. code-block:: go
            
            package main
            import "fmt"

            func main() {
               fmt.Println("Hello World!")
            }
                        

>          You'd typically compile this via `go build -o app app.go`.

Reviewer: ```suggestion
         You'd typically compile this via ``go build -o app app.go``.
```

   
	.. tab:: Rust

>          Assume the following `app.rs` source code:

Reviewer: ```suggestion
         Assume the following ``app.rs`` source code:
```

         .. code-block:: rust
      
            
            fn main() {

>             println!("Hello World!");

Reviewer: ```suggestion
               println!("Hello World!");
```

            }

               

>          You'd typically compile this via `rustc app.rs`.

Reviewer: ```suggestion
         You'd typically compile this via ``rustc app.rs``.
```

> To build the final application image, you simply need to add your compiled executable to the `chiselled-ubuntu-base:latest` container image. So your new Dockerfile should be similar to:

Reviewer: ```suggestion
To build the final application image, you simply need to add your compiled executable to the ``chiselled-ubuntu-base:latest`` container image. So your new Dockerfile should be similar to:
```

.. code-block:: dockerfile

   FROM chiselled-ubuntu-base:latest
   COPY app /
   ENTRYPOINT [ "./app" ]

> Build this chiselled application image with `docker build  -t chiselled-app:latest .` and then run it:

Reviewer: ```suggestion
Build this chiselled application image with ``docker build  -t chiselled-app:latest .`` and then run it:
```
